File: src-tauri/src/source_pdf.rs
use std::fs;
use std::path::PathBuf;
use tauri::{AppHandle, Manager};

fn paper_dir(app: &AppHandle, paper_id: &str) -> Result<PathBuf, String> {
    let root = app
        .path()
        .app_data_dir()
        .map_err(|e| e.to_string())?
        .join("papers")
        .join(paper_id);
    Ok(root)
}

#[tauri::command]
pub fn save_source_pdf(app: AppHandle, paper_id: String, data: Vec<u8>) -> Result<String, String> {
    let dir = paper_dir(&app, &paper_id)?;
    fs::create_dir_all(&dir).map_err(|e| e.to_string())?;
    let path = dir.join("source.pdf");
    fs::write(&path, &data).map_err(|e| e.to_string())?;
    Ok(path.to_string_lossy().to_string())
}

#[tauri::command]
pub fn copy_source_pdf(
    app: AppHandle,
    paper_id: String,
    source_path: String,
) -> Result<String, String> {
    let dir = paper_dir(&app, &paper_id)?;
    fs::create_dir_all(&dir).map_err(|e| e.to_string())?;
    let dest = dir.join("source.pdf");
    fs::copy(&source_path, &dest).map_err(|e| e.to_string())?;
    Ok(dest.to_string_lossy().to_string())
}

#[tauri::command]
pub fn open_source_pdf(app: AppHandle, paper_id: String, page: Option<u32>) -> Result<(), String> {
    let path = paper_dir(&app, &paper_id)?.join("source.pdf");
    if !path.exists() {
        return Err("保存されたPDFが見つかりません".into());
    }
    let _ = page;
    #[cfg(target_os = "macos")]
    {
        std::process::Command::new("open")
            .arg(&path)
            .spawn()
            .map_err(|e| e.to_string())?;
        return Ok(());
    }
    #[cfg(not(target_os = "macos"))]
    {
        Err("このOSでは元PDFを開く操作に未対応です".into())
    }
}

#[tauri::command]
pub fn delete_source_pdf(app: AppHandle, paper_id: String) -> Result<(), String> {
    let dir = paper_dir(&app, &paper_id)?;
    if dir.exists() {
        fs::remove_dir_all(&dir).map_err(|e| e.to_string())?;
    }
    Ok(())
}

#[tauri::command]
pub fn source_pdf_exists(app: AppHandle, paper_id: String) -> Result<bool, String> {
    Ok(paper_dir(&app, &paper_id)?.join("source.pdf").exists())
}

File: src/services/sourcePdf.ts
import { isTauriApp } from "../utils/serverReady";
import type { BoundingBox } from "../types/paper";

export type OpenSourcePdfArgs = {
  paperId: string;
  page?: number;
  boundingBox?: BoundingBox;
};

function nativePath(file: File): string | undefined {
  const withPath = file as File & { path?: string };
  return typeof withPath.path === "string" && withPath.path.length > 0
    ? withPath.path
    : undefined;
}

/**
 * Copy the original PDF into Application Support.
 * Browser-only `npm run dev` cannot persist to the app data dir.
 */
export async function persistSourcePdf(
  paperId: string,
  file: File
): Promise<string | null> {
  if (!isTauriApp()) return null;
  const { invoke } = await import("@tauri-apps/api/core");
  const path = nativePath(file);
  if (path) {
    return invoke<string>("copy_source_pdf", {
      paperId,
      sourcePath: path,
    });
  }
  const bytes = new Uint8Array(await file.arrayBuffer());
  return invoke<string>("save_source_pdf", {
    paperId,
    data: Array.from(bytes),
  });
}

export async function openSourcePdf(args: OpenSourcePdfArgs): Promise<void> {
  if (!isTauriApp()) {
    throw new Error("元PDFはデスクトップアプリから開けます");
  }
  const { invoke } = await import("@tauri-apps/api/core");
  await invoke("open_source_pdf", {
    paperId: args.paperId,
    page: args.page ?? null,
  });
}

export async function deleteStoredSourcePdf(paperId: string): Promise<void> {
  if (!isTauriApp()) return;
  const { invoke } = await import("@tauri-apps/api/core");
  await invoke("delete_source_pdf", { paperId });
}

export async function sourcePdfExists(paperId: string): Promise<boolean> {
  try {
    const { getStorage } = await import("../data/runtime");
    const { fs } = await getStorage();
    if (await fs.exists(`papers/${paperId}/source.pdf`)) return true;
  } catch {
    // fall through to the Tauri command
  }
  if (!isTauriApp()) return false;
  const { invoke } = await import("@tauri-apps/api/core");
  return invoke<boolean>("source_pdf_exists", { paperId });
}
